Neos.Flow/Classes/Mvc/Routing/Dto/RouteTags.php

```diff
@@ -0,0 +1,145 @@
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\Flow\Annotations as Flow;
+
+/**
+ * RouteTags to be associated with matched/resolved routes
+ *
+ * RouteTags can be added by Route Part handlers via the ResolveResult/MatchResult return values
+ * The tags will be added to the corresponding cache entries, so that they can be flushed selectively using the RouterCachingService
+ *
+ * @Flow\Proxy(false)
+ */
+final class RouteTags
+{
+
+    /**
+     * Pattern a tag must match. @see \Neos\Cache\Frontend\FrontendInterface::PATTERN_TAG
+     */
+    const PATTERN_TAG = '/^[a-zA-Z0-9_%\-&]{1,250}$/';
+
+    /**
+     * @var string[] numeric array of strings satisfying the PATTERN_TAG regex
+     */
+    private $tags = [];
+
+    /**
+     * @param string[] $tags numeric array of strings satisfying the PATTERN_TAG regex
+     */
+    private function __construct(array $tags)
+    {
+        $this->tags = $tags;
+    }
+
+    /**
+     * Creates an instance without any tags
+     *
+     * @return RouteTags
+     */
+    public static function createEmpty(): self
+    {
+        return new static([]);
+    }
+
+    /**
+     * Creates an instance with one given tag
+     *
+     * @param string $tag Tag value satisfying the PATTERN_TAG regex
+     * @return RouteTags
+     */
+    public static function createFromTag(string $tag): self
+    {
+        self::validateTag($tag);
+        return new static([$tag]);
+    }
+
+    /**
+     * Creates an instance with one given tags
+     *
+     * @param string[] $tags An array of strings satisfying the PATTERN_TAG regex
+     * @return RouteTags
+     */
+    public static function createFromArray(array $tags): self
+    {
+        array_walk($tags, 'static::validateTag');
+        return new static($tags);
+    }
+
+    /**
+     * Merges two instances of this class combining and unifying all tags
+     *
+     * @param RouteTags $tags
+     * @return RouteTags
+     */
+    public function merge(RouteTags $tags): self
+    {
+        $mergedTags = array_unique(array_merge($this->tags, $tags->tags));
+        return new static($mergedTags);
+    }
+
+    /**
+     * Creates a new instance with the given $tag added
+     * If the $tag has been added already, this instance is returned
+     *
+     * @param string $tag
+     * @return RouteTags
+     */
+    public function withTag(string $tag): self
+    {
+        if ($this->has($tag)) {
+            return $this;
+        }
+        self::validateTag($tag);
+        $newTags = $this->tags;
+        $newTags[] = $tag;
+        return new static($newTags);
+    }
+
+    /**
+     * Checks the format of a given $tag string and throws an exception if it does not conform to the PATTERN_TAG regex
+     *
+     * @param string $tag
+     * @throws \InvalidArgumentException
+     */
+    private static function validateTag($tag)
```

Comment:
RouteTags look like a collection, why not creating RouteTag object to encapsulate the validation logic ?

Reply:
True. Validation should be done in a tag object, not in tags. Then again, no blocker for me, as it's private and can be changed any time later…

Reply:
I agree, that a tag would make a good Value Object and in fact in an earlier version I had that. But if we go that path we should introduce a generic `Tag` value object IMO, maybe in a package `neos/tags` and refer to that from all places where we currently use strings for tags (like in `neos/cache`).
I could kickstart that project to be included with the next Flow/Neos versions but as Karsten wrote it won't really change the API of this one